gcovr/workers.py

```diff
@@ -57,6 +57,22 @@ def locked_directory(dir_):
 locked_directory.global_object = LockedDirectories()
 
 
+def drain(queue):
+    """
+    Drain a queue of all items
+    """
+    sentinel_count = 0
+    while True:
+        try:
+            work, args, kwargs = queue.get(False)
```

Comment:
Is it correct to use nonblocking operations here? I think that would make it possible that items are added after this loop completes and before sentinels are enqueued.

Instead, it might be better to block until we get a sentinel? Or could that lead to deadlocks, which are worse than a few extra items? Or should we enqueue a sentinel before we start draining it. Or should we introduce a condition variable that controls whether the queue is open? That condvar would then be set as soon as an exception is received.

Reply:
I'm not sure about this myself. I was aiming for simplicity with a possibility of some extra work in the case of error. Perhaps abstracting the error mechanism into the Workers class so that the worker asks the Workers class to register an exception. At that point the Workers class could put locks around adding work to the queue and raising errors + draining the queue.

I think if we don't allow an addition at the same time as raising an exception - and we use a separate mutex to the queue - we shouldn't get deadlock.

Reply:
I think maybe the `if not exceptions` test in `add()` actually serves as a kind of mutex and prevents enqueueing of extra items, if the `exceptions` is appended before draining.

In any case – you are the multithreading expert here. If you think you have a better solution it might be worth a shot.

At this point, it may be best to extract all queue operations (add, sentinel, dequeue, drain) into a separate class so that it's easier to see what is going on. But that can also be done outside of this PR.

Reply:
I've moved things into the pool and supplied a lock. I agree it makes things easier to read to have the methods split up and it means that the mutex is explicit and checked properly.